tests/unit/lib/test_runner/test_template_generator.py

```diff
@@ -0,0 +1,347 @@
+            "                  PolicyDocument:\n",
+            "                      Version: 2012-10-17\n",
+            "                      Statement:\n",
+            "                          - Sid: S3BucketAccess\n",
+            "                            Effect: Allow\n",
+            "                            Action:\n",
+            "                                - s3:PutObject\n",
+            "                                - s3:GetObject\n",
+            "                            Resource: !Sub\n",
+            "                                - arn:aws:s3:::${bucket}/*\n",
+            "                                - bucket: !Ref S3Bucket\n",
+            "                        # Generated IAM Actions #\n",
+        ]
+        self.generated_template_expected_second_half = [
+            "\n",
+            "    TaskDefinition:\n",
+            "        Type: AWS::ECS::TaskDefinition\n",
+            "        Properties:\n",
+            "            RequiresCompatibilities:\n",
+            "                - FARGATE\n",
+            "            ExecutionRoleArn: !Sub arn:aws:iam::${AWS::AccountId}:role/ecsTaskExecutionRole\n",
+            "            TaskRoleArn: !Ref ContainerIAMRole\n",
+            "            Cpu: 256\n",
+            "            Memory: 512\n",
+            "            NetworkMode: awsvpc\n",
+            "            ContainerDefinitions:\n",
+            "                - Name: cloud-test-python-container\n",
+            "                  Image: 123456789123.dkr.ecr.us-east-1.amazonaws.com/cloud-test-repo\n",
+            "                  PortMappings:\n",
+            "                      - ContainerPort: 8080\n",
+            "                        Protocol: tcp\n",
+            "                  LogConfiguration:\n",
+            "                      LogDriver: awslogs\n",
+            "                      Options:\n",
+            "                          awslogs-region: !Ref AWS::Region\n",
+            "                          awslogs-group: !Ref LogGroup\n",
+            "                          awslogs-stream-prefix: ecs\n",
+            "\n",
+            "    LogGroup:\n",
+            "        Type: AWS::Logs::LogGroup\n",
+            "        Properties:\n",
+            "            LogGroupName: cloud-test-loggroup\n",
+            "\n",
+            "    ECSCluster:\n",
+            "        Type: AWS::ECS::Cluster\n",
+            "        Properties:\n",
+            "            ClusterName: cloud-test-fargate-cluster\n",
+            "\n",
+            "    S3Bucket:\n",
+            "        Type: AWS::S3::Bucket\n",
+            "        Properties:\n",
+            "            BucketName: cloud-test-bucket-unique-name",
+        ]
+
+    @patch("samcli.lib.test_runner.test_runner_template_generator._query_tagging_api")
+    def test_lambda_function(self, query_tagging_api_patch):
```

Comment:
Most of the test_<resource type> functions are nearly identical; it is probably worth factoring out the common core such that each test function sets the resource arn and expected statements and the common code constructs the expected and actual results, then compares them.

Reply:
Will do!